src/gsi/hybrid_ensemble_isotropic.F90

```diff
@@ -3833,46 +3927,84 @@ subroutine ckgcov_a_en_new_factorization(ig,z,a_en)
   else
 ! Apply horizontal smoother for number of horizontal scales
      if(regional) then
+        if(l_mgbf_loc) then
+           ii=0
+           do k=1,obj_mgbf(ig)%km_all
+              do j=1-obj_mgbf(ig)%hy,obj_mgbf(ig)%jm+obj_mgbf(ig)%hy
+                 do i=1-obj_mgbf(ig)%hx,obj_mgbf(ig)%im+obj_mgbf(ig)%hx
+                    ii=ii+1
+                    obj_mgbf(ig)%VALL(k,i,j)=z(ii)
+                 enddo
+              enddo
+           enddo
+           call obj_mgbf(ig)%filtering_procedure(obj_mgbf(ig)%mgbf_proc,1)
+        else
 ! Make a copy of input variable z to hwork
-        hwork=z
-        iadvance=2 ; iback=1
-        call new_factorization_rf_y(hwork,iadvance,iback,grd_loc%kend_loc+1-grd_loc%kbegin_loc,ig)
-        call new_factorization_rf_x(hwork,iadvance,iback,grd_loc%kend_loc+1-grd_loc%kbegin_loc,ig)
+           hwork=z
+           iadvance=2 ; iback=1
+           call new_factorization_rf_y(hwork,iadvance,iback,grd_loc%kend_loc+1-grd_loc%kbegin_loc,ig)
+           call new_factorization_rf_x(hwork,iadvance,iback,grd_loc%kend_loc+1-grd_loc%kbegin_loc,ig)
+! Put back onto subdomains
+           call general_grid2sub(grd_loc,hwork,a_en_work)
+        endif
      else
 #ifdef LATER
         call sqrt_sf_xy(ig,z,hwork,grd_loc%kbegin_loc,grd_loc%kend_loc)
 #else
         write(6,*) ' problem with ibm compiler with "use hybrid_ensemble_isotropic, only: sqrt_sf_xy"'
 #endif /*LATER*/
+! Put back onto subdomains
+        call general_grid2sub(grd_loc,hwork,a_en_work)
      end if
   end if
 
-! Put back onto subdomains
-  allocate(a_en_work(n_ens*a_en(1)%ndim),stat=istatus)
-  if(istatus/=0) then
-     write(6,*)'ckgcov_a_en_new_factorization: trouble in alloc(a_en_work)'
-     call stop2(999)
-  endif
-  call general_grid2sub(grd_loc,hwork,a_en_work)
-
 ! Retrieve ensemble components from long vector
-  ii=0
-  do k=1,n_ens
-     is=ii+1
-     ie=ii+a_en(1)%ndim
-     a_en(k)%values(1:a_en(k)%ndim)=a_en_work(is:ie)
-     ii=ii+a_en(1)%ndim
-  enddo
-  deallocate(a_en_work)
-
 ! Apply vertical smoother on each ensemble member
-  iadvance=2 ; iback=1
-!$omp parallel do schedule(static,1) private(k)
-  do k=1,n_ens
-
-     call new_factorization_rf_z(a_en(k)%r3(ipnt)%q,iadvance,iback,ig)
-
-  enddo
+  if(regional.and.l_mgbf_loc) then
+     allocate(work_mgbf(obj_mgbf(ig)%km_a_all,obj_mgbf(ig)%nm,obj_mgbf(ig)%mm))
+     work_mgbf=zero
+     if(obj_mgbf(ig)%km_a_all==obj_mgbf(ig)%km_all &
```

Comment:
That will be better if those branches in mgbf could be done in mgbf lib , namely using a generic mgbf call wrapping up the different options here.

Reply:
That is one possible method, but I don't think MGBF lib should depend on GSI, so mapping between a_en (GSI working array) and work_mgbf (MGBF working array) were excluded from MGBF lib and done in GSI.

Reply:
@shoyokota  those options are specific to mgbf , right?   If so, wrapping them into mgbf doesn't add extra dependency on gsi, in fact, it makes mgbf more independent of GSI.

Reply:
Now I understand you are talking only about the part of mapping between analysis and filter grids. Yes, you are right. I wrapped this part as subroutines "anal_to_filt_allmap" and "filt_to_anal_allmap" in mgbf lib.